workflows/pipe-common/scripts/autoscale_sge.py

```diff
@@ -799,97 +813,156 @@ class MemoryHostStorage:
+        self._storage = dict()
+        self.clock = Clock()
 
     def add_host(self, host):
         if host in self._storage:
             raise ScalingError('Host with name \'%s\' is already in the host storage' % host)
-        self._storage.append(host)
+        self._storage[host] = self.clock.now()
 
     def remove_host(self, host):
-        if host not in self._storage:
-            raise ScalingError('Host with name \'%s\' doesn\'t exist in the host storage' % host)
-        self._storage.remove(host)
+        self._validate_existence(host)
+        self._storage.pop(host)
+
+    def update_hosts_activity(self, hosts, timestamp):
+        for host in hosts:
+            self._validate_existence(host)
+            self._storage[host] = timestamp
+
+    def get_hosts_activity(self, hosts):
+        hosts_activity = {}
+        for host in hosts:
+            self._validate_existence(host)
+            hosts_activity[host] = self._storage[host]
+        return hosts_activity
 
     def load_hosts(self):
-        return list(self._storage)
+        return list(self._storage.keys())
 
     def clear(self):
-        self._storage = list()
+        self._storage = dict()
+
+    def _validate_existence(self, host):
+        if host not in self._storage:
+            raise ScalingError('Host with name \'%s\' doesn\'t exist in the host storage' % host)
 
 
 class FileSystemHostStorage:
     _REPLACE_FILE = 'echo "%(content)s" > %(file)s_MODIFIED; ' \
                     'mv %(file)s_MODIFIED %(file)s'
+    _DATETIME_FORMAT = '%m/%d/%Y %H:%M:%S'
+    _VALUE_BREAKER = '|'
     _LINE_BREAKER = '\n'
 
-    def __init__(self, cmd_executor, storage_file):
+    def __init__(self, cmd_executor, storage_file, clock=Clock(), master_host=None):
         """
         Additional hosts storage.
-
+        Contains the hostname along with the time of the last activity on it.
         It uses file system to persist all hosts. Therefore it can be used through relaunches of the autoscaler script.
 
         :param cmd_executor: Cmd executor.
         :param storage_file: File to store hosts into.
         """
         self.executor = cmd_executor
         self.storage_file = storage_file
+        self.clock = clock
+        self.master_host = master_host
 
     def add_host(self, host):
         """
         Persist host to storage.
 
         :param host: Additional host name.
         """
-        hosts = self.load_hosts()
+        hosts = self._load_hosts_stats()
         if host in hosts:
             raise ScalingError('Host with name \'%s\' is already in the host storage' % host)
-        hosts.append(host)
+        hosts[host] = self.clock.now()
         self._update_storage_file(hosts)
 
+    def update_hosts_activity(self, hosts, timestamp):
+        latest_hosts_stats = self._load_hosts_stats()
+        for host in hosts:
+            if host != self.master_host:
+                self._validate_existence(host, latest_hosts_stats)
```

Comment:
We can get rid of this and all other additional `master_host` management since master host won't be written to the hosts  storage in the first place.

Reply:
Done